### DISCUSSION TITLE
How can I enable SQLCMD mode?

### DISCUSSION THREAD

**Discussion by martingbrown** [microsoft/azuredatastudio#24216] [Q&A] [answered]
Hello,

I've read that there should be an SQLCMD mode to allow us to edit SQLCMD scripts. In the [September 2019 release notes](https://cloudblogs.microsoft.com/sqlserver/2019/09/10/the-september-release-of-azure-data-studio-is-now-available/) Alan Yu suggested there should be a button at the top of the SQL editor pane to turn this on and off. I'm running the 1.45.1 (user setup) version and I don't seem to have this button. Has anyone any idea how I get it?

Martin

**Comment by erinstellato-ms** [comment]
@martingbrown Screenshot from a query editor window, if you don't have the same, could you post a screenshot?  Thanks!

![image](https://github.com/microsoft/azuredatastudio/assets/98599126/ffd7010c-a502-4eab-a60d-22c56d540d1f)

**Reply by martingbrown** [reply]
I definitely don't have that button.

**Reply by martingbrown** [reply]
![image](https://github.com/microsoft/azuredatastudio/assets/5264795/85483255-0290-4240-a07b-fe1d2aa35f89)

**Comment by erinstellato-ms** [comment, marked as answer]
@martingbrown I learned something new today, apparently we have that behind the preview setting.  If you go into Settings ( CTRL/CMD + , ) and then type Enable Preview Features, you'll see the option to enable:

![image](https://github.com/microsoft/azuredatastudio/assets/98599126/8dfff222-b5ab-4257-b2e6-b90b2ceb512b)

One you enable that and re-open the query editor, you should see the option.

I'll open an issue to enable the option without the preview flag.  Thanks for bringing this to our attention!

**Reply by martingbrown** [reply]
Yes, thats done it. Thank you. I thought it would be something like that. Though I'm a little surprised as this was in the release notes three years ago give or take a couple of weeks, seems to have been in preview a long time.

**Reply by erinstellato-ms** [reply]
@martingbrown Agreed!  Opened a PR and assuming I'm not missing anything, it will get into Insiders by Monday or Tuesday and be in the next stable release.

**Reply by erinstellato-ms** [reply]
@martingbrown Follow up, I did some research to understand why it's still behind the preview flag.  Currently, we don't support all commands, and from my testing, we don't have syntax highlighting yet.  So, we are keeping this behind the flag for now.  I'm not sure how quickly we can make the necessary improvements to make it generally available, based on the other projects already on the roadmap.  But please log issues if you find them and if this is critical to your daily workflow, please note that and encourage friends/colleagues to upvote any issues as that helps us prioritize requests.

**Reply by martingbrown** [reply]
I was just writing some PowerShell scripts that need to interact with a database. I'm using Invoke-SqlCmd to run queries and needed to pass parameters. As long as it doesn't mark $(VAR) as an error I'm happy.
